python/pyiceberg/utils/config.py

```diff
@@ -0,0 +1,139 @@
+
+
+def _coalesce(lhs: Optional[Any], rhs: Optional[Any]) -> Optional[Any]:
+    return lhs or rhs
+
+
+def _merge_config(lhs: RecursiveDict, rhs: RecursiveDict) -> RecursiveDict:
+    """merges right-hand side into the left-hand side"""
+    new_config = lhs.copy()
+    for rhs_key, rhs_value in rhs.items():
+        if rhs_key in new_config:
+            lhs_value = new_config[rhs_key]
+            if isinstance(lhs_value, dict) and isinstance(rhs_value, dict):
+                # If they are both dicts, then we have to go deeper
+                new_config[rhs_key] = _merge_config(lhs_value, rhs_value)
+            else:
+                # Take the non-null value, with precedence on rhs
+                new_config[rhs_key] = _coalesce(rhs_value, lhs_value)
+        else:
+            # New key
+            new_config[rhs_key] = rhs_value
+
+    return new_config
+
+
+def _lowercase_dictionary_keys(input_dict: RecursiveDict) -> RecursiveDict:
+    """Lowers all the keys of a dictionary in a recursive manner, to make the lookup case-insensitive"""
+    return {k.lower(): _lowercase_dictionary_keys(v) if isinstance(v, dict) else v for k, v in input_dict.items()}
+
+
+class Config:
+    config: RecursiveDict
+
+    def __init__(self):
+        config = self._from_configuration_files() or {}
+        config = _merge_config(config, self._from_environment_variables(config))
+        self.config = FrozenDict(**config)
+
+    @staticmethod
+    def _from_configuration_files() -> Optional[RecursiveDict]:
+        """Looks up configuration files
+
+        Takes preference over RecursiveDict that's close to the user
+        """
+        for directory in [os.environ.get(HOME), os.environ.get(PYICEBERG_HOME)]:
+            if directory:
+                path = f"{directory.rstrip('/')}/{PYICEBERG_YML}"
+                if os.path.isfile(path):
+                    with open(path, encoding="utf-8") as f:
+                        try:
+                            file_config = yaml.safe_load(f)
+                            file_config_lowercase = _lowercase_dictionary_keys(file_config)
+                            return file_config_lowercase
+                        except yaml.YAMLError:
+                            logger.exception("Not valid config yaml: %s", path)
+        return None
+
+    @staticmethod
+    def _from_environment_variables(config: RecursiveDict) -> RecursiveDict:
+        """Reads the environment variables, to check if there are any prepended by PYICEBERG_
+
+        Args:
+            config: Existing configuration that's being amended with configuration from environment variables
+
+        Returns:
+            Amended configuration
+        """
+
+        def set_property(_config: RecursiveDict, path: List[str], config_value: str):
+            while len(path) > 0:
+                element = path.pop(0)
+                if element not in _config:
+                    _config[element] = {}
+                if len(path) == 0:
```

Comment:
Minor: if last element is missing it will set the config to `{}` before setting it to `config_value`.

Reply:
Nice one, thanks!